File: src/main/java/com/implimentz/fragments/Fragment.java
package com.implimentz.fragments;

import android.content.res.Configuration;
import android.support.annotation.CallSuper;
import android.support.annotation.MenuRes;
import android.support.annotation.StringRes;
import android.support.v4.content.ContextCompat;
import android.support.v7.app.AppCompatActivity;
import android.support.v7.view.ActionMode;
import android.support.v7.widget.Toolbar;
import android.view.LayoutInflater;
import android.view.Menu;
import android.view.MenuItem;
import android.view.View;
import android.view.ViewGroup;
import android.widget.AdapterView;
import android.widget.ImageView;

/**
 * Created by Alexander Efremenkov.
 * Date: 05.01.16, 11:10
 * In Intellij IDEA 15.0.1 Ultimate
 * email: implimentz@gmail.com
 * twitter: iamironz
 */
@SuppressWarnings("unchecked")
public class Fragment<D, A extends AppCompatActivity> {

    private boolean isFinished = false;
    private boolean isShowing = false;
    private boolean configurationChanged = false;

    private View view;
    private D args;
    private A activity;

    private String title;
    private String subTitle;
    private ActionMode actionMode;
    private Toolbar toolbar;

    public Fragment() {

    }

    public Fragment(final D args) {
        this.args = args;
    }

    final void setActivity(final A owner) {
        this.activity = owner;
    }

    final void clearViews() {
        if (view != null) {
            clearViews(view);
        }
    }

    private void clearViews(final View v) {
        if (v == null) {
            return;
        }
        try {
            if (v.getBackground() != null) {
                v.getBackground().setCallback(null);
                return;
            }
            if (v instanceof ImageView) {
                ((ImageView) v).setImageBitmap(null);
                return;
            }
            if (v instanceof ViewGroup && !(v instanceof AdapterView)) {
                for (int i = 0; i < ((ViewGroup) v).getChildCount(); i++) {
                    clearViews(((ViewGroup) v).getChildAt(i));
                }
                ((ViewGroup) v).removeAllViews();
            }
        } catch (Exception ignored) {
        }
    }

    final void setConfigurationChanged() {
        configurationChanged = true;
    }

    public void onConfigurationChanged(final Configuration newConfig) {

    }

    public void onLowMemory() {

    }

    final View constructView(final ViewGroup container) {
        if (view == null || configurationChanged) {
            view = onCreateView(activity.getLayoutInflater(), container, args);
        }

        configurationChanged = false;
        isShowing = true;
        return view;
    }

    public View onCreateView(final LayoutInflater inflater, final ViewGroup container, final D args) {
        return new View(activity);
    }

    @CallSuper
    public void onResume() {
        isShowing = true;
    }

    @CallSuper
    public void onPause() {
        isShowing = false;
    }

    @CallSuper
    public void onDestroy() {
        isFinished = true;
        isShowing = false;
        configurationChanged = false;
        clearViews();
        view = null;
        args = null;
        activity = null;
        title = null;
        subTitle = null;
        actionMode = null;
        toolbar = null;
    }

    public boolean hasNotEndedAction() {
        return false;
    }

    public void onActionEndRequired() {

    }

    public final ActionMode startActionMode(final ActionMode.Callback callback) {
        actionMode = activity.startSupportActionMode(callback);
        return actionMode;
    }

    public final void finishActionMode() {
        if (actionMode == null) {
            return;
        }

        actionMode.finish();
    }

    public boolean onActionModeEnabled() {
        return false;
    }

    @MenuRes
    public int onCreateOptionMenu(final Menu menu) {
        return 0;
    }

    @CallSuper
    public boolean onOptionsItemSelected(final MenuItem item) {
        return false;
    }

    public final boolean isFinished() {
        return isFinished;
    }

    public final boolean isRunned() {
        return !isFinished;
    }

    public final boolean isShowing() {
        return isShowing;
    }

    public final boolean isHidden() {
        return !isShowing;
    }

    public final View getView() {
        return view;
    }

    public final D getArgs() {
        return args;
    }

    public final A getActivity() {
        return activity;
    }

    public final String getString(@StringRes final int stringId) {
        return getActivity().getString(stringId);
    }

    public final int getColor(final int colorId) {
        return ContextCompat.getColor(activity, colorId);
    }

    final String getTitle() {
        return title;
    }

    final String getSubTitle() {
        return subTitle;
    }

    final void setToolbar(final Toolbar toolbar) {
        this.toolbar = toolbar;
    }

    public final void onBackPressed() {
        activity.onBackPressed();
    }

    public final void setTitle(final String title) {
        if (isShowing && !isFinished) {
            toolbar.setTitle(title);
        }
        this.title = title;
    }

    public final void setSubTitle(final String subTitle) {
        if (isShowing && !isFinished) {
            toolbar.setSubtitle(subTitle);
        }
        this.subTitle = subTitle;
    }
}

File: src/main/java/com/implimentz/fragments/Fragment.kt
@file:Suppress("unused")

package com.implimentz.fragments

import android.content.res.Configuration
import android.support.annotation.CallSuper
import android.view.LayoutInflater
import android.view.MenuItem
import android.view.View
import android.view.ViewGroup
import com.implimentz.fragments.annotation.AnnotationManager
import java.io.Serializable

/**
 * Created by ironz.
 * Date: 19.01.16, 17:59
 * In Intellij IDEA 15.0.3 Ultimate
 * email: implimentz@gmail.com
 * twitter: iamironz
 */

/**
 * Base class for using as fragment container that represent activity lifecycle with specific lifecycle methods.
 *
 * @param arguments (optional) - for passing parametrized argument with some specific data needed by fragment.
 * Passing for [arguments] return parametrized argument and getting into [onViewCreated] method.
 * Must implement [Serializable] marker interface or if argument not passing set generic type as [NoArgs].
 * Please make sure to call super() constructor.
 */
open class Fragment<Data : Serializable>(val arguments: Data? = null) {

    @Volatile
    internal var configurationChanged: Boolean = false
    @Volatile
    var finished: Boolean = false
        private set
    @Volatile
    var showing: Boolean = false
        private set
    @Volatile
    var name: String? = null
        internal set
    @Volatile
    var view: View? = null
        private set

    /**
     * Function that calls by App [android.content.ComponentCallbacks] while screen orientation is changed
     *
     * @param newConfig - updated application config
     */
    @CallSuper
    open fun onConfigurationChanged(newConfig: Configuration) {

    }

    /**
     * Function that calls while fragment manager need to get new\created view
     *
     * @param container - target container for showing content in
     * @param inflater - activity inflater for inflate [com.implimentz.fragments.annotation.LayoutMeta]
     */
    internal fun constructView(container: ViewGroup, inflater: LayoutInflater): View {
        tryRecreateView(container, inflater)
        updateStartFlags()
        return view as View
    }

    /**
     * Function that creates view while view is null or device config has been changed
     *
     * @param container - target container for showing content in
     * @param inflater - activity inflater for inflate [com.implimentz.fragments.annotation.LayoutMeta]
     */
    private fun tryRecreateView(container: ViewGroup, inflater: LayoutInflater) {
        if ((view == null) or configurationChanged) {
            view = inflateViewFromLayoutRes(container, inflater)
            onViewCreated(view as View, arguments)
        }
    }

    /**
     * inflate layout from annotation or throws exception if annotation was not found
     * @see [AnnotationManager.getLayoutOrThrow]
     *
     * @param container - target container for showing content in
     * @param inflater - activity inflater for inflate [com.implimentz.fragments.annotation.LayoutMeta]
     */
    private fun inflateViewFromLayoutRes(container: ViewGroup, inflater: LayoutInflater): View {
        val meta = AnnotationManager.getLayoutOrThrow(this)
        return inflater.inflate(meta.value, container, false)
    }

    /**
     * Method that calls after @FragmentLayout layout created and added in container
     *
     * @param view - created view using for data binding
     * @param arguments - argument that was passed in fragment constructor
     */
    @CallSuper
    open fun onViewCreated(view: View, arguments: Data?) {

    }

    /**
     * Method that calls after container activity or fragment created/restored
     */
    @CallSuper
    open fun onResume() {
        showing = true
    }

    /**
     * Method that calls after container activity or fragment is minimized/overlapped
     */
    @CallSuper
    open fun onPause() {
        showing = false
    }

    /**
     * Method that calls after activity configuration changed/destroyed
     */
    @CallSuper
    open fun onDestroy() {
        updateStopFlags()
    }

    /**
     * Method that calls for update as active fragment status flags
     */
    private fun updateStartFlags() {
        finished = false
        showing = true
        configurationChanged = false
    }

    /**
     * Method that calls for update as inactive fragment status flags
     */
    private fun updateStopFlags() {
        finished = true
        showing = false
        configurationChanged = false
    }

    /**
     * Method that calls after FragmentManager onBackPressed() method called
     */
    open fun hasNotEndedAction(): Boolean {
        return false
    }

    /**
     * Method that calls after hasNotEndedAction() has return 'true'
     */
    open fun onActionEndRequired() {

    }

    /**
     * Method that calls after some menu item was clicked
     */
    @CallSuper
    open fun onOptionsItemSelected(item: MenuItem) {

    }

    /**
     * Method that calls after back button was pressed
     */
    open fun onBackPressed() {

    }
}
